graph/path/yen_ksp_test.go

```diff
@@ -0,0 +1,113 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package path
+
+import (
+	"math"
+	"reflect"
+	"testing"
+
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/simple"
+)
+
+var yenShortestPathTests = []struct {
+	name  string
+	graph func() graph.WeightedEdgeAdder
+	edges []simple.WeightedEdge
+
+	query     simple.Edge
+	k         int
+	wantPaths [][]int64
+}{
```

Comment:
I'd like to see tests that have k > |paths|, tests on a G={V={0,1}, E={(0,1)}}, on a G={V={0}, E=∅}, on a G=star_n, and on a G={V={-1,1,2..n}, E={(-1,m),(m,1)}} for m∈[2,n] and s=-1, t=1. for k<|{m}|, k = |{m}| and k>|{m}|.

Reply:
I have a star n graph test cases, but I don't quite get the second test case you are asking for.

Reply:
I'd like to abandon this PR in favour of #525.

The bipartite graph tests look at a variety of cases where I was expecting the implementation to break down (the k > |paths| too, which I altered in #525 where you had k = |paths|). The algorithm as stated (?) and implemented attempts to give back k paths, but does this with such force that it gives back k possibly paths, in the case that there are fewer paths than k, some returned paths are ∅ and some do not contain t. This looks like a failure to me. It does not look like there is a test in the algorithm at WP for this case, but it is fairly trivial to add. Those changes have been made in the new PR, along with some additional docs and a minor optimisation for allocations. Note that the bipartite graph tests can be extended to include non-equally weighted s-t paths, which is probably worth doing.

Please feel free to comment at #525, but I also want input from others, since, on the face of it, the algorithm seems to have defects and what I've done feels like papering over them.